### PR TITLE
Mal implementation by TypeScript

### PR DESCRIPTION
**Pull request by vvakame**
Hi! I tried to implement mal by [TypeScript](http://www.typescriptlang.org/).
It was very fun! thanks 😉 

This PR checked by above script.

```
$ make "test^ts"
# PASSED!
$ make MAL_IMPL=ts "test^mal"
# PASSED!
```

I can't run `make DOCKERIZE=1 "test^ts"` locally.
Is there a way to exec the local image?

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -856,6 +856,17 @@ cd tcl
 tclsh ./stepX_YYY.tcl
 ```
 
+### TypeScript
```
Comment (dubek) [inline README.md]: Please add a credit line to yourself! (see example in the TCL implementation just above).

**Comment by rhysd** [comment]
@vvakame you should increment the number of languages and add TypeScript to the language list at the top of README :+1:

**Comment by kanaka** [comment]
@vvakame in addition to the changes suggested by @dubek and @rhysd, the ts/Makefile should have rules for each final target. That way when tests are invoked from the top-level, the needed files will be automatically built. Here is a Makefile that works for me:

```
SOURCES_BASE = types.ts reader.ts printer.ts
SOURCES_LISP = env.ts core.ts stepA_mal.ts
SOURCES = $(SOURCES_BASE) $(SOURCES_LISP)

STEPS = step0_repl step1_read_print step2_eval step3_env \
        step4_if_fn_do step5_tco step6_file step7_quote \
        step8_macros step9_try stepA_mal

all: ts

node_modules:
        npm install

step%.js: node_modules
        ./node_modules/.bin/tsc -p ./


.PHONY: ts clean stats tests $(TESTS)

ts: $(foreach s,$(STEPS),$(s).js)

clean:
        rm -f *.js

stats: $(SOURCES)
        @wc $^
        @printf "%5s %5s %5s %s\n" `grep -E "^[[:space:]]*//|^[[:space:]]*$$" $^ | wc` "[comments/blanks]"
stats-lisp: $(SOURCES_LISP)
        @wc $^
        @printf "%5s %5s %5s %s\n" `grep -E "^[[:space:]]*//|^[[:space:]]*$$" $^ | wc` "[comments/blanks]"
```

I have also built and pushed a docker image based on your Dockerfile. You should now be able to update `.travis.yml` with TypeScript and the next time you update this pull request it should test your implementation. Once you've made those changes and Travis passes, then I'll merge it. Also, feel free to join #mal on freenode. It's mostly quiet with occasional discussion and announcements related to mal/miniMAL/educational and esoteric Lisps.

**Comment by vvakame** [comment]
thanks all!
I update this PR.
Travis CI passed 😸

**Comment by kanaka** [comment]
@vvakame Excellent! Merged.

**Comment by kanaka** [comment]
@vvakame only remaining thing is for you to tweet to announce it and I'll retweet. Or if you prefer, I can tweet it.

**Comment by vvakame** [comment]
https://twitter.com/vvakame/status/836010542456160257 😺

**Pull request merged**